Show adjusted AR/OD when using DT/HT

Givikap120 opened the pull request:
When you're using the DT on AR9 map, it shows AR9 in map info.
Now it will show true AR, 10.33

Givikap120 committed:
Added a true AR/OD display when using DT/HT

bdach commented on osu.Game/Screens/Select/Details/AdvancedStats.cs:
```diff
@@ -126,6 +126,22 @@ private void updateStatistics()
+                {
+                    double speedChange = (float)mod.SpeedChange.Value;
+
+                    double preempt = (int)IBeatmapDifficultyInfo.DifficultyRange(adjustedDifficulty.ApproachRate, 1800, 1200, 450) / speedChange;
+                    adjustedDifficulty.ApproachRate = (float)(preempt > 1200 ? (1800 - preempt) / 120 : (1200 - preempt) / 150 + 5);
```
not sure how to feel about this being hardcoded here like that. it's essentially specific to osu!-the-ruleset, catch _completely coincidentally_ uses the same formula. probably should be moved out to rulesets or something

stanriders replied:
it might fail on custom rulesets so it should totally be moved out

bdach commented on osu.Game/Screens/Select/Details/AdvancedStats.cs:
```diff
@@ -126,6 +126,22 @@ private void updateStatistics()
+                    double preempt = (int)IBeatmapDifficultyInfo.DifficultyRange(adjustedDifficulty.ApproachRate, 1800, 1200, 450) / speedChange;
+                    adjustedDifficulty.ApproachRate = (float)(preempt > 1200 ? (1800 - preempt) / 120 : (1200 - preempt) / 150 + 5);
+
+                    float hitwindow300 = (80.0f - 6 * adjustedDifficulty.OverallDifficulty) / (float)speedChange;
+                    adjustedDifficulty.OverallDifficulty = (80.0f - hitwindow300) / 6;
```
this is however 100% incorrect for anything else but osu!-the-ruleset (mania OD is not even going to be affected by rate adjusting mods very soon, see https://github.com/ppy/osu/issues/24603).

Givikap120 committed:
Fixed "works only for std" problem

1) Now it use AR/OD calculation from Ruleset class
2) Implemented needed functions in each of default rulesets

Givikap120 committed:
Moved ruleset decl in right place

Givikap120 committed:
Added "value is rate changed" indication

Now if AR or OD value is not "true" and changed through rate - it will appear with symbol `*`
Instead of `11` it wil show as `11*`

bdach commented on osu.Game/Rulesets/Ruleset.cs:
```diff
@@ -389,5 +389,13 @@ public IEnumerable<(HitResult result, LocalisableString displayName)> GetHitResu
         /// </summary>
         public virtual DifficultySection? CreateEditorDifficultySection() => null;
+
+        public virtual double PreemptFromAr(float AR) => 0;
+        public virtual float ArFromPreempt(double preempt) => 5;
+        public virtual double HitwindowFromOd(float OD) => 0;
+        public virtual float OdFromHitwindow(double hitwindow300) => 0;
+
+        public virtual float ChangeArFromRate(float AR, double rate) => AR;
+        public virtual float ChangeOdFromRate(float OD, double rate) => OD;
```
No. This is not acceptable. What happened with [yesterday's proposal](https://discord.com/channels/188630481301012481/1097318920991559880/1144338816400900198) of having `public virtual BeatmapDifficulty GetEffectiveDifficulty()`?

To explain why this is not acceptable: rulesets can be and do all sorts of things. How do you define this if a ruleset has no concept of "approach rate"? How do you define this if a ruleset has no concept of "hit windows" (catch)?

Givikap120 replied:
This is made for conveniency. I thought it will be useful to use one function instead of using inline implementations of "AR to preempt" kinda things. 
The main functions here are `ChangeArFromRate` and `ChangeOdFromRate`.

About the `GetEffectiveDifficulty()`.
```
isArRateAdjusted = false;
isOdRateAdjusted = false;
```
Those two lines are made to indicate that AR/OD changed thru rate (and not "true"). This kinda goes against the idea of `GetEffectiveDifficulty()` that just returns effective difficulty.

Givikap120 committed:
fixed stated problems

removed functions from `Ruleset.cs` and added a `GetEffectiveDifficulty()` instead

Givikap120 committed:
Merge pull request #1 from Givikap120/master

add display to calibrate

Givikap120 committed:
Added BeatmapAttributesDisplay support

peppy commented on osu.Game/Screens/Select/Details/AdvancedStats.cs:
```diff
@@ -223,6 +221,7 @@ public LocalisableString Title
                     bar.Length = value.baseValue / maxValue;
 
                     valueText.Text = (value.adjustedValue ?? value.baseValue).ToString(forceDecimalPlaces ? "0.00" : "0.##");
+                    if (value.isRateAdjusted) valueText.Text += "*";
```
Ignoring the code quality of this PR (I am not ready to even attempt to figure it out), this method of display is ambiguous. A user would have little idea what this means.

peppy commented:
I'd recommend discussing the UI/UX of this before implementing. The code here is not usable from a code quality / structural perspective, so I'd rather we're on the same page visually before fixing anything.

Can you show visually what you are going for here so we can start a discussion on that first?

- Adding a "*" after a number with no tooltip or explanation isn't something I'd want to do
- If we're showing "effective" AR, the original should also be shown next to it, with the effective value displayed alongside it somehow.

Givikap120 commented:
> I'd recommend discussing the UI/UX of this before implementing. The code here is not usable from a code quality / structural perspective, so I'd rather we're on the same page visually before fixing anything.
> 
> Can you show visually what you are going for here so we can start a discussion on that first?
> 
> * Adding a "*" after a number with no tooltip or explanation isn't something I'd want to do
> * If we're showing "effective" AR, the original should probably be shown next to it, with the effective value displayed alongside it somehow.

What is this thing currently doing:
![image](https://github.com/ppy/osu/assets/89256026/890a335e-086c-4574-b096-bf6029ddf735)
![image](https://github.com/ppy/osu/assets/89256026/d5e22b83-6244-4b50-9b0d-714112d878f2)
![image](https://github.com/ppy/osu/assets/89256026/1e67b69d-aafd-41a2-90b1-0ca7d977f257)
![image](https://github.com/ppy/osu/assets/89256026/16b83554-5619-4e83-92a1-789245cc9e59)
![image](https://github.com/ppy/osu/assets/89256026/d30402d3-149a-4eb8-b0cb-2262ce6b55e8)

Joehuu commented:
Rate changing mods also affect [HP drain](https://osu.ppy.sh/wiki/en/Beatmap/HP_drain_rate#mod-effects), so it should also be visualized even though the effective value is not mentioned/used that much.

Givikap120 commented:
> Rate changing mods also affect [HP drain](https://osu.ppy.sh/wiki/en/Beatmap/HP_drain_rate#mod-effects), so it should also be visualized even though the effective value is not mentioned/used that much.

I'm don't sure that this is any useful. No matter: DT or NM, same misses result in same hp loss, while increasing hp make you loose more hp for same miss count. 
HP is draining faster, but notes to recover it appears faster as well, so it's compensate any potential difficulty increase.
It doesn't work like this with AR/OD. AR becoming clearly higher (with slightly different animation tho) and OD hitwindow becoming clearly narrower without any compensation.

peppy commented:
If we're showing AR changes, then HP should be too. It's the same deal.

Givikap120 commented:
> If we're showing AR changes, then HP should be too. It's the same deal. I think you're very confused saying "it's different to AR / OD". It's 100% the same.

ok, I need some clarification
if you will speedup song by 1.5x using smth like osu trainer - will it be easier to pass than DT one?
and if it's - how exactly? it's the same as changing HP to value that makes HP drain 1.5x faster?

peppy commented:
> how exactly? it's the same as changing HP to value that makes HP drain 1.5x faster?

that's the same as AR and OD. the only thing *really* changing is time scale. think about it for a moment.

Givikap120 commented:
> If we're showing AR changes, then HP should be too. It's the same deal. I think you're very confused saying "it's different to AR / OD". It's 100% the same.

I'm judging from the perspective of DT being alternative to just manually shrink all the notes and edit the mp3
If you will do this - AR and OD will remain the same, so you need to adjust them. But HP drain rate will increase without touching HP parameter (I've tested it), because unlike AR and OD, HP is also dependent on notes and their time position.

peppy commented:
>I'm judging from the perspective of DT being alternative to just manually shrink all the notes and edit the mp3
If you will do this - AR and OD will remain the same, so you need to adjust them. But HP drain rate will increase without touching HP parameter (I've tested it), because unlike AR and OD, HP is also dependent on notes and their time position.

Maybe it's a perspective thing, but I can't wrap my head around this way of thinking.

I guess one reason showing HP change would be hard is that it isn't really a linear scale (at least in stable's implementation).

Givikap120 commented:
> > I'm judging from the perspective of DT being alternative to just manually shrink all the notes and edit the mp3
> > If you will do this - AR and OD will remain the same, so you need to adjust them. But HP drain rate will increase without touching HP parameter (I've tested it), because unlike AR and OD, HP is also dependent on notes and their time position.
> 
> Maybe it's a perspective thing, but I can't wrap my head around this way of thinking.
> 
> I guess one reason showing HP change would be hard is that it isn't really a linear scale (at least in stable's implementation).

I propose mark HP as yellow (neutral) to mark that it's somewhat behave differently, but without changing the actual number

Givikap120 commented:
also, what about making a vertical line hover-button on the right
when a user hovers over it - the values change to "not-rate-change" and colors returns to white

Givikap120 commented:
and this hover-button will have a tool tip that tells about values being rate-changed

peppy commented:
If this is to be merged, there definitely needs to be tooltips which allow viewing the original values, as proposed.

Givikap120 commented:
> If this is to be merged, there definitely needs to be tooltips which allow viewing the original values, as proposed.

My idea is this:
![image](https://github.com/ppy/osu/assets/89256026/1fd35857-e8a1-48c1-86b7-e520e84e20ee)

After hovering over a button:
![image](https://github.com/ppy/osu/assets/89256026/f2aae9c3-63ab-4b1e-ad5f-6836d77242a9)

After hovering long enough:
![image](https://github.com/ppy/osu/assets/89256026/15718b70-2d36-4aa1-8467-165de14a06cd)

peppy commented:
Let's not make new UI elements and just add tooltips for now please.

Givikap120 commented:
> Let's not make new UI elements and just add tooltips for now please.

so how you will show the original values then?
in tooltip?

peppy commented:
I feel like I've said that twice already.

Just simple text like "Without mods: {starRating}" would be ample.

Givikap120 committed:
Improve code quality

Givikap120 committed:
Basic rate-adjust tooltip

Givikap120 committed:
fixed bug in AR formula

Givikap120 committed:
improved tooltip

Givikap120 committed:
Update CatchRuleset.cs

Givikap120 committed:
Update AdvancedStats.cs

Givikap120 commented:
> I feel like I've said that twice already.
> 
> Just simple text like "Without mods: {starRating}" would be ample.

okay, I'm sorry, I'm blind
review the PR again please

peppy commented on osu.Game.Rulesets.Osu/OsuRuleset.cs:
```diff
@@ -328,5 +328,21 @@ public override StatisticItem[] CreateStatisticsForScore(ScoreInfo score, IBeatm
 
         public override RulesetSetupSection CreateEditorSetupSection() => new OsuSetupSection();
+
+        public double PreemptFromAr(float AR) => AR < 5 ? (1200.0 + 600.0 * (5 - AR) / 5) : (1200.0 - 750.0 * (AR - 5) / 5);
+        public float ArFromPreempt(double preempt) => (float)(preempt > 1200 ? ((1800 - preempt) / 120) : ((1200 - preempt) / 150 + 5));
+        public double HitwindowFromOd(float OD) => 80.0 - 6 * OD;
+        public float OdFromHitwindow(double hitwindow300) => (float)(80.0 - hitwindow300) / 6;
+        public float ChangeArFromRate(float AR, double rate) => ArFromPreempt(PreemptFromAr(AR) / rate);
+        public float ChangeOdFromRate(float OD, double rate) => OdFromHitwindow(HitwindowFromOd(OD) / rate);
```
What on earth is this mess..

Givikap120 replied:
I wanted to put all these functions that are usually just copied to different places into one place.
I shouldn't do this? I can put all of this code inside the function `GetRateAdjustedDifficulty`

peppy replied:
> I wanted to put all these functions that are usually just copied to different places into one place.

You may have identified other usages, but you haven't updated any of them to use these new functions. Ignoring the fact that these functions have no xmldoc and are named badly, it doesn't make sense to do this unless you also centralise all existing usages.

If you're going to do this, please do it in a separate PR as a prerequisite or follow-up to this change. For now I'd recommend inlining these.

Givikap120 committed:
Delete util functions from rulesets

peppy commented on osu.Game.Rulesets.Mania/ManiaRuleset.cs:
```diff
@@ -427,6 +427,20 @@ public override IRulesetFilterCriteria CreateRulesetFilterCriteria()
 
         public override DifficultySection CreateEditorDifficultySection() => new ManiaDifficultySection();
+
+        // Mania doesn't have rate-adjusted attributes anymore?
+
+        //public override BeatmapDifficulty GetRateAdjustedDifficulty(IBeatmapDifficultyInfo baseDifficulty, double rate)
+        //{
+        //    BeatmapDifficulty adjustedDifficulty = new BeatmapDifficulty(baseDifficulty);
+
+        //    double hitwindow = 64.0 - 3 * adjustedDifficulty.OverallDifficulty;
+        //    hitwindow /= rate;
+        //    adjustedDifficulty.OverallDifficulty = (float)(64.0 - hitwindow) / 3;
+
+        //    return adjustedDifficulty ?? (BeatmapDifficulty)baseDifficulty;
+        //}
+
```
Shouldn't exist.

Givikap120 replied:
okay, I've just wanted to overask about it

Givikap120 replied:
fixed

peppy commented on osu.Game/Rulesets/Ruleset.cs:
```diff
@@ -389,5 +389,6 @@ public IEnumerable<(HitResult result, LocalisableString displayName)> GetHitResu
         /// Can be overridden to alter the difficulty section to the editor beatmap setup screen.
         /// </summary>
         public virtual DifficultySection? CreateEditorDifficultySection() => null;
+        public virtual BeatmapDifficulty GetRateAdjustedDifficulty(IBeatmapDifficultyInfo baseDifficulty, double rate) => new BeatmapDifficulty(baseDifficulty);
```
xmldoc

Givikap120 replied:
added

Givikap120 committed:
FIxed stated problems

Givikap120 committed:
Update Ruleset.cs

Givikap120 committed:
Update ManiaRuleset.cs

Givikap120 committed:
fixed naming incconvinence

Givikap120 commented:
Please, recheck again

peppy committed:
Merge branch 'master' into arod_rate_adjust

peppy committed:
Refactor logic and tooltip formatting

peppy committed:
Apply various inspections

Givikap120 commented:
![image](https://github.com/ppy/osu/assets/89256026/fcc41664-ba79-407c-b326-3c03567193c7)
I will fix the shown precision issue

Givikap120 committed:
Fixed precision and updated AdvancedStats

peppy committed:
Add note about localisation

peppy committed:
Rename method and adjust xmldoc to be very explicit about how wrong this is

peppy commented:
The tooltip code can't be copy paste spewed across two classes. Should be converted to a shared custom tooltip before merge.

And I'm not even sure if/when this will be mergeable because of all the `Ruleset` functions. I know for one I don't have the time to go through and confirm all these local implementations match expectations. Dunno.

bdach commented:
I've personally long given up on this PR and only see it getting in if I basically rewrite it.

Givikap120 commented:
> The tooltip code can't be copy paste spewed across two classes. Should be converted to a shared custom tooltip before merge.
> 
> And I'm not even sure if/when this will be mergeable because of all the `Ruleset` functions. I know for one I don't have the time to go through and confirm all these local implementations match expectations. Dunno.

Where should this function be if not in a ruleset?
It's literally ruleset-specific behavior because each ruleset has its own formulas.

I'll try to make a custom tooltip tho, but I don't know how it's even possible without adding those functions in the Ruleset class.

peppy commented:
> Where should this function be if not in a ruleset?

I'm not complaining about the location; I'm complaining about local implementations of mathematical equations likely used elsewhere in the codebase, with little information / reference to help assert they are correct.

Givikap120 commented:
> > Where should this function be if not in a ruleset?
> 
> I'm not complaining about the location; I'm complaining about local implementations of mathematical equations likely used elsewhere in the codebase, with little information / reference to help assert they are correct.

I've taken them from the osu!wiki, how I can make them clearer?
Add a comment that is copied text from osu wiki?
Also, for example, AR to preempt is here, but preempt to AR isn't mentioned in wiki, so I've just used simple math.

Givikap120 committed:
Made custom tooltip

Givikap120 commented:
Ok, is this design suitable? Or should I make something more complex?
![image](https://github.com/ppy/osu/assets/89256026/f9f7d1f7-7746-4ddd-90a7-393efe659b0e)
![image](https://github.com/ppy/osu/assets/89256026/c7dac457-4257-4930-8886-8d24ecc8d1cd)

Givikap120 committed:
slight format changes

peppy commented:
That format definitely looks better than what we had.

Givikap120 commented:
> That format definitely looks better than what we had.

so is this good and we only need quality of code changes?
or I should try to make something better than this?

Givikap120 committed:
fixed test errors

Givikap120 committed:
Update BeatmapAttributesDisplay.cs

Givikap120 committed:
Update AdjustedAttributesTooltip.cs

Givikap120 committed:
Merge branch 'master' into arod_rate_adjust

Givikap120 committed:
fixed one test

peppy committed:
Merge branch 'master' into arod_rate_adjust

peppy committed:
Update in line with upstream changes

peppy committed:
Fix various code quality and visual issues with `AdjustedAttributesTooltip`

peppy committed:
Tidy up `GetRateAdjustedDisplayDifficulty` implemetations

peppy committed:
Refactor everything for sanity

peppy commented:
@Givikap120 please check the changes I made to get this merged. arguably i had to rewrite the whole tooltip changes you made for simplicity. there was no real reason to add the bindable flow that you did, so i've nuked it all.

Givikap120 commented:
> @Givikap120 please check the changes I made to get this merged. arguably i had to rewrite the whole tooltip changes you made for simplicity. there was no real reason to add the bindable flow that you did, so i've nuked it all.

Well, I've made this bindable flow because I thought that clearing and re-adding attribute containers was bad, so I just updated existing ones with bindables.
